composable_kernel/include/tensor_operation/gridwise_gemm_pipeline_v1.hpp

```diff
@@ -0,0 +1,245 @@
+            b_blockwise_copy.RunRead(b_grid_desc, b_grid_buf);
+
+            // Initialize C
+            c_thread_buf.Clear();
+
+            a_blockwise_copy.RunWrite(a_block_desc, a_block_buf);
+            b_blockwise_copy.RunWrite(b_block_desc, b_block_buf);
+
+            // main body
+            if constexpr(HasMainLoop)
+            {
+                index_t i = 0;
+
+                do
+                {
+                    a_blockwise_copy.MoveSrcSliceWindow(a_grid_desc, a_block_copy_step);
+                    b_blockwise_copy.MoveSrcSliceWindow(b_grid_desc, b_block_copy_step);
+
+                    a_blockwise_copy.RunRead(a_grid_desc, a_grid_buf);
+
+                    block_sync_lds();
+
+                    b_blockwise_copy.RunRead(b_grid_desc, b_grid_buf);
+
+                    blockwise_gemm.Run(a_block_buf, b_block_buf, c_thread_buf);
+
+                    block_sync_lds();
+
+                    a_blockwise_copy.RunWrite(a_block_desc, a_block_buf);
+                    b_blockwise_copy.RunWrite(b_block_desc, b_block_buf);
+
+                    ++i;
+                } while(i < (NumLoop - 1));
+            }
+
+            // tail
+            {
+                block_sync_lds();
+
+                blockwise_gemm.Run(a_block_buf, b_block_buf, c_thread_buf);
+            }
+#else
+            // preload data into LDS
+            a_blockwise_copy.RunRead(a_grid_desc, a_grid_buf);
+            b_blockwise_copy.RunRead(b_grid_desc, b_grid_buf);
+
+            a_blockwise_copy.MoveSrcSliceWindow(a_grid_desc, a_block_copy_step);
+            b_blockwise_copy.MoveSrcSliceWindow(b_grid_desc, b_block_copy_step);
+
+            // Initialize C
+            c_thread_buf.Clear();
+
+            a_blockwise_copy.RunWrite(a_block_desc, a_block_buf);
+            b_blockwise_copy.RunWrite(b_block_desc, b_block_buf);
+
+            // main body
+            if constexpr(HasMainLoop)
+            {
+                index_t i = 0;
+
+                do
+                {
+                    a_blockwise_copy.RunRead(a_grid_desc, a_grid_buf);
+
+                    block_sync_lds();
+
+                    b_blockwise_copy.RunRead(b_grid_desc, b_grid_buf);
+
+                    blockwise_gemm.Run(a_block_buf, b_block_buf, c_thread_buf);
+
+                    block_sync_lds();
+
+                    a_blockwise_copy.MoveSrcSliceWindow(a_grid_desc, a_block_copy_step);
+                    b_blockwise_copy.MoveSrcSliceWindow(b_grid_desc, b_block_copy_step);
+
+                    a_blockwise_copy.RunWrite(a_block_desc, a_block_buf);
+                    b_blockwise_copy.RunWrite(b_block_desc, b_block_buf);
+
+                    ++i;
+                } while(i < (NumLoop - 1));
+            }
+
+            // tail
+            {
+                block_sync_lds();
+
+                blockwise_gemm.Run(a_block_buf, b_block_buf, c_thread_buf);
+            }
+#endif
+        }
+        else if constexpr(NumPrefetch == 2)
```

Comment:
I'd preferably move these different pipelines to their own template specializations. Things like:

```c++
template <
    typename Params,
    index_t NumPrefetch>
struct GridwiseGemmPipeline_v1; // class template declaration

template <typename Params>
struct GridwiseGemmPipeline_v1<Params, 1>
{
     // implements 1 stage prefetch
}

template <typename Params>
struct GridwiseGemmPipeline_v1<Params, 2>
{
     // implements 2 stage prefetch
}
```

The upside of this approach is that when someone accidentally calls nonexistent 3 stage prefetch it triggers compilation error immediately instead of silently compiles. But of course you already have `static_assert` to guard against exactly the case. So, just matter of personal preference.

Reply:
fixed